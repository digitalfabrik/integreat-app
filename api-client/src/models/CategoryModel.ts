import { Moment } from 'moment'

import ExtendedPageModel from './ExtendedPageModel'
import PageModel from './PageModel'

class CategoryModel extends ExtendedPageModel {
  _root: boolean
  _parentPath: string
  _order: number

  constructor(params: {
    root: boolean
    path: string
    title: string
    content: string
    thumbnail: string
    parentPath: string
    order: number
    availableLanguages: Map<string, string>
    lastUpdate: Moment
    hash: string
  }) {
    const { order, parentPath, root, ...other } = params
    super(other)
    this._root = root
    this._parentPath = parentPath
    this._order = order
  }

  get parentPath(): string {
    return this._parentPath
  }

  get order(): number {
    return this._order
  }

  isRoot(): boolean {
    return this._root
  }

  isEqual(other: PageModel): boolean {
    return (
      other instanceof CategoryModel &&
      super.isEqual(other) &&
      this.parentPath === other.parentPath &&
      this.order === other.order &&
      this.isRoot === other.isRoot
    )
  }
}

export default CategoryModel
