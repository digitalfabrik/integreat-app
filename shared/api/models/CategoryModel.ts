import { DateTime } from 'luxon'

import normalizePath from '../../utils/normalizePath'
import ExtendedPageModel from './ExtendedPageModel'
import OfferModel from './OfferModel'
import OrganizationModel from './OrganizationModel'
import PageModel from './PageModel'

class CategoryModel extends ExtendedPageModel {
  _root: boolean
  _parentPath: string
  _order: number
  _organization: OrganizationModel | null
  _embeddedOffers: OfferModel[]

  constructor(params: {
    root: boolean
    path: string
    title: string
    content: string
    thumbnail: string | null
    parentPath: string
    order: number
    availableLanguages: Map<string, string>
    lastUpdate: DateTime
    organization: OrganizationModel | null
    embeddedOffers: OfferModel[]
  }) {
    const { order, parentPath, root, organization, embeddedOffers, ...other } = params
    super(other)
    this._root = root
    this._parentPath = normalizePath(parentPath)
    this._order = order
    this._organization = organization
    this._embeddedOffers = embeddedOffers
  }

  get embeddedOffers(): OfferModel[] {
    return this._embeddedOffers
  }

  get parentPath(): string {
    return this._parentPath
  }

  get order(): number {
    return this._order
  }

  isRoot(): boolean {
    return this._root
  }

  get organization(): OrganizationModel | null {
    return this._organization
  }

  isEqual(other: PageModel): boolean {
    return (
      other instanceof CategoryModel &&
      super.isEqual(other) &&
      this.parentPath === other.parentPath &&
      this.order === other.order &&
      this.isRoot === other.isRoot &&
      (this.organization === null ? other.organization === null : this.organization.isEqual(other.organization)) &&
      this.embeddedOffers.length === other.embeddedOffers.length &&
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      this.embeddedOffers.every((offer, index) => offer.isEqual(other.embeddedOffers[index]!))
    )
  }
}

export default CategoryModel
