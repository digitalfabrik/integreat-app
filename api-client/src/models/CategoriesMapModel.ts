import normalizePath from 'normalize-path'

import CategoryModel from './CategoryModel'

/**
 * Contains a Map [string -> CategoryModel] and some helper functions
 */

class CategoriesMapModel {
  _categories: Map<string, CategoryModel>

  /**
   * Creates a Map [path -> category] from the categories provided,
   * whose parent attributes are first changed from id to path
   * @param categories CategoryModel as array
   */
  constructor(categories: Array<CategoryModel>) {
    this._categories = new Map(categories.map(category => [category.path, category]))
  }

  /**
   * @return {CategoryModel[]} categories The categories as array
   */
  toArray(): Array<CategoryModel> {
    return Array.from(this._categories.values())
  }

  /**
   * Returns the category with the given path
   * @param {String} path The path
   * @return {CategoryModel | undefined} The category
   */
  findCategoryByPath(path: string): CategoryModel | null | undefined {
    return this._categories.get(decodeURIComponent(normalizePath(path)))
  }

  /**
   * Returns all children of the given category
   * @param category The category
   * @return {CategoryModel[]} The children
   */
  getChildren(category: CategoryModel): Array<CategoryModel> {
    return this.toArray()
      .filter(_category => _category.parentPath === category.path)
      .sort((category1, category2) => category1.order - category2.order)
  }

  /**
   * Returns all (mediate) parents of the given category
   * @param category The category
   * @return {CategoryModel[]} The parents, with the immediate parent last
   */
  getAncestors(category: CategoryModel): Array<CategoryModel> {
    const parents: Array<CategoryModel> = []
    let currentCategory = category

    while (!currentCategory.isRoot()) {
      const temp = this.findCategoryByPath(currentCategory.parentPath)

      if (!temp) {
        throw new Error(
          `The category ${currentCategory.parentPath} does not exist but should be the parent of ${currentCategory.path}`
        )
      }

      currentCategory = temp
      parents.unshift(currentCategory)
    }

    return parents
  }

  isLeaf(category: CategoryModel): boolean {
    return this.getChildren(category).length === 0
  }

  isEqual(other: CategoriesMapModel): boolean {
    return (
      this._categories.size === other._categories.size &&
      Array.from(this._categories.entries()).every(([key, value]) => {
        const otherCategory = other._categories.get(key)

        return otherCategory && value.isEqual(otherCategory)
      })
    )
  }
}

export default CategoriesMapModel
