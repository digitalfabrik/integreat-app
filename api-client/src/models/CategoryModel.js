// @flow

import type Moment from 'moment'
import CategoriesMapModel from './CategoriesMapModel'
import ExtendedPageModel from './ExtendedPageModel'

class CategoryModel extends ExtendedPageModel {
  _root: boolean
  _parentPath: string
  _order: number

  constructor (params: {|
    root: boolean, path: string, title: string, content: string, thumbnail: string,
    parentPath: string, order: number, availableLanguages: Map<string, string>, lastUpdate: Moment, hash: string
  |}) {
    const { order, parentPath, root, ...other } = params
    super(other)
    this._root = root
    this._parentPath = parentPath
    this._order = order
  }

  get parentPath (): string {
    return this._parentPath
  }

  get order (): number {
    return this._order
  }

  isRoot (): boolean {
    return this._root
  }

  isLeaf (categories: CategoriesMapModel): boolean {
    return categories.getChildren(this).length === 0
  }
}

export default CategoryModel
