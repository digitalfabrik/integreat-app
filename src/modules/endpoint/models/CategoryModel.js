// @flow

import type Moment from 'moment'
import CategoriesMapModel from './CategoriesMapModel'

class CategoryModel {
  _id: number
  _path: string
  _title: string
  _content: string
  _parentPath: string
  _thumbnail: string
  _order: number
  _availableLanguages: Map<string, string>
  _lastUpdate: ?Moment

  constructor (params: {| id: number, path: string, title: string, content: string, thumbnail: string,
    parentPath: string, order: number, availableLanguages: Map<string, string>, lastUpdate: ?Moment |}) {
    this._id = params.id
    this._path = params.path
    this._title = params.title
    this._content = params.content
    this._parentPath = params.parentPath
    this._thumbnail = params.thumbnail
    this._order = params.order
    this._availableLanguages = params.availableLanguages
    this._lastUpdate = params.lastUpdate
  }

  get thumbnail (): string {
    return this._thumbnail
  }

  get id (): number {
    return this._id
  }

  get path (): string {
    return this._path
  }

  get title (): string {
    return this._title
  }

  get content (): string {
    return this._content
  }

  get parentPath (): string {
    return this._parentPath
  }

  get order (): number {
    return this._order
  }

  get availableLanguages (): Map<string, string> {
    return this._availableLanguages
  }

  get lastUpdate (): ?Moment {
    return this._lastUpdate
  }

  isRoot (): boolean {
    return this.id === 0
  }

  isLeaf (categories: CategoriesMapModel): boolean {
    return categories.getChildren(this).length === 0
  }
}

export default CategoryModel
