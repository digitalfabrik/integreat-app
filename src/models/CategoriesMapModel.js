// @flow

import normalizePath from 'normalize-path'
import CategoryModel from './CategoryModel'

/**
 * Contains a Map [string -> CategoryModel] and some helper functions
 */
class CategoriesMapModel {
  _categories: Map<string, CategoryModel>
  /**
   * Creates a Map [path -> category] from the categories provided,
   * whose parent attributes are first changed from id to path
   * @param categories CategoryModel as array
   */
  constructor (categories: Array<CategoryModel>) {
    this._categories = new Map(categories.map(category => ([category.path, category])))
  }

  /**
   * @return {CategoryModel[]} categories The categories as array
   */
  toArray (): Array<CategoryModel> {
    return Array.from(this._categories.values())
  }

  /**
   * Returns the category with the given path
   * @param {String} path The path
   * @return {CategoryModel | undefined} The category
   */
  findCategoryByPath (path: string): ?CategoryModel {
    return this._categories.get(decodeURIComponent(normalizePath(path)))
  }

  /**
   * Returns all children of the given category
   * @param category The category
   * @return {CategoryModel[]} The children
   */
  getChildren (category: CategoryModel): Array<CategoryModel> {
    return this.toArray()
      .filter(_category => _category.parentPath === category.path)
      .sort((category1, category2) => (category1.order - category2.order))
  }

  /**
   * Returns all (mediate) parents of the given category
   * @param category The category
   * @return {CategoryModel[]} The parents, with the immediate parent last
   */
  getAncestors (category: CategoryModel): Array<CategoryModel> {
    const parents = []

    while (!category.isRoot()) {
      const temp = this.findCategoryByPath(category.parentPath)
      if (!temp) {
        throw new Error(`The category ${category.parentPath} does not exist but should be the parent of ${category.path}`)
      }
      category = temp
      parents.unshift(category)
    }
    return parents
  }
}

export default CategoriesMapModel
