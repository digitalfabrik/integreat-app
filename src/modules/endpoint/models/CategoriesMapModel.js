// @flow

import normalizeUrl from 'normalize-url'
import CategoryModel from './CategoryModel'

/**
 * Contains a Map [string -> CategoryModel] and some helper functions
 */
class CategoriesMapModel {
  _categories: Map<string, CategoryModel>
  /**
   * Creates a Map [url -> category] from the categories provided,
   * whose parent attributes are first changed from id to url
   * @param categories CategoryModel as array
   */
  constructor (categories: Array<CategoryModel>) {
    this._categories = new Map(categories.map(category => ([category.url, category])))
  }

  /**
   * @return {CategoryModel[]} categories The categories as array
   */
  toArray (): Array<CategoryModel> {
    return Array.from(this._categories.values())
  }

  /**
   * Returns the category with the given url
   * @param {String} url The url
   * @return {CategoryModel} The category
   */
  getCategoryByUrl (url: string): CategoryModel {
    const category = this._categories.get(normalizeUrl(url))
    if (!category) {
      throw Error(`No category with the given url '${url}'`)
    }
    return category
  }

  /**
   * Returns the category with the given id
   * @param id The id
   * @return {CategoryModel | undefined} The category
   */
  getCategoryById (id: number): ?CategoryModel {
    const category = this.toArray().find(category => category.id === id)

    if (!category) {
      throw Error(`No category with the given id '${id}'`)
    }
    return category
  }

  /**
   * Returns all children of the given category
   * @param category The category
   * @return {CategoryModel[]} The children
   */
  getChildren (category: CategoryModel): Array<CategoryModel> {
    return this.toArray()
      .filter(_category => _category.parentUrl === category.url)
      .sort((category1, category2) => (category1.order - category2.order))
  }

  /**
   * Returns all (mediate) parents of the given category
   * @param category The category
   * @return {CategoryModel[]} The parents, with the immediate parent last
   */
  getAncestors (category: CategoryModel): Array<CategoryModel> {
    const parents = []

    while (category.parentUrl !== '') {
      category = this.getCategoryByUrl(category.parentUrl)
      parents.unshift(category)
    }
    return parents
  }
}

export default CategoriesMapModel
