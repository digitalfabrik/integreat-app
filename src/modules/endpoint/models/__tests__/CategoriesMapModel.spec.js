import CategoriesMapModel from '../CategoriesMapModel'
import CategoryModel from '../CategoryModel'
import moment from 'moment'

describe('CategoriesMapModel', () => {
  const farsiCategory = new CategoryModel({
    id: 404,
    path: '/augsburg/fa/erste-schritte/نقشه-شهر',
    title: 'نقشه شهر',
    content: '',
    parentPath: '/augsburg/fa/erste-schritte',
    thumbnail: 'https://cms.integreat-app.de/augsburg/wp-content/uploads/sites/2/2015/09/pin66-150x150.png',
    order: 3,
    availableLanguages: new Map([
      [
        'ar',
        '/augsburg/ar/erste-schritte/خريطة-المدينة'
      ]
    ]),
    lastUpdate: moment('2016-01-07 10:36:24')
  })

  const categories = [
    new CategoryModel({id: 0, path: '/augsburg/de', title: 'augsburg', parentPath: ''}),
    new CategoryModel({
      id: 20,
      path: '/augsburg/de/willkommen',
      parentPath: '/augsburg/de',
      title: 'willkommen',
      order: 1
    }),
    new CategoryModel({
      id: 21,
      path: '/augsburg/de/erste-schritte',
      parentPath: '/augsburg/de',
      title: 'erste-schritte',
      order: 2
    }),
    new CategoryModel({
      id: 22,
      path: '/augsburg/de/erste-schritte/asylantrag',
      parentPath: '/augsburg/de/erste-schritte',
      title: 'asylantrag'
    })
  ]

  const categoriesMapModel = new CategoriesMapModel(categories)

  const category1 = categoriesMapModel.findCategoryByPath('/augsburg/de/')
  const category2 = categoriesMapModel.findCategoryByPath('/augsburg/de/willkommen')
  const category3 = categoriesMapModel.findCategoryByPath('/augsburg/de/erste-schritte///')
  const category4 = categoriesMapModel.findCategoryByPath('/augsburg/de/erste-schritte/asylantrag')

  it('should get the right categories and normalize url components', () => {
    expect(category1).toEqual(categories[0])
    expect(category2).toEqual(categories[1])
    expect(category3).toEqual(categories[2])
    expect(category4).toEqual(categories[3])
  })

  it('should normalize url components correctly for special characters', () => {
    const map = new CategoriesMapModel([farsiCategory])
    const foundFarsiCategory = map.findCategoryByPath('/augsburg/fa/erste-schritte/%D9%86%D9%82%D8%B4%D9%87-%D8%B4%D9%87%D8%B1')
    expect(foundFarsiCategory).toEqual(farsiCategory)
  })

  it('should have the right parent attributes', () => {
    expect(category2.parentPath).toBe(category1.path)
    expect(category3.parentPath).toBe(category1.path)
    expect(category4.parentPath).toBe(category3.path)
  })

  it('should return all (mediate) parents in right order', () => {
    expect(categoriesMapModel.getAncestors(category4)[0]).toEqual(category1)
    expect(categoriesMapModel.getAncestors(category4)[1]).toEqual(category3)
  })

  it('should return all immediate children in the right order', () => {
    expect(categoriesMapModel.getChildren(category1)[0]).toEqual(category2)
    expect(categoriesMapModel.getChildren(category1)[1]).toEqual(category3)
  })

  it('should instruct CategoryModel to be leaf', () => {
    expect(category4.isLeaf(categoriesMapModel)).toBe(true)
    expect(category3.isLeaf(categoriesMapModel)).toBe(false)
  })
})
