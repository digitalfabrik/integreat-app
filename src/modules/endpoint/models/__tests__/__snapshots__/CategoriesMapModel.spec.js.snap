// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesMapModel should find category by id 1`] = `"No category with the given id '1234567'"`;

exports[`CategoriesMapModel should get the right categories and normalize urls 1`] = `"No category with the given url '/test/url'"`;
